staging/src/k8s.io/client-go/tools/cache/shared_informer.go

```diff
@@ -660,18 +736,30 @@ func (p *sharedProcessor) run(stopCh <-chan struct{}) {
 	func() {
 		p.listenersLock.RLock()
 		defer p.listenersLock.RUnlock()
-		for _, listener := range p.listeners {
+		for listener := range p.listeners {
 			p.wg.Start(listener.run)
 			p.wg.Start(listener.pop)
 		}
 		p.listenersStarted = true
 	}()
 	<-stopCh
-	p.listenersLock.RLock()
-	defer p.listenersLock.RUnlock()
-	for _, listener := range p.listeners {
-		close(listener.addCh) // Tell .pop() to stop. .pop() will tell .run() to stop
-	}
+
+	func() {
```

Comment:
What was the problem with holding lock through the wg.Wait?  Holding the lock until the shutdown was complete prevented any other thread from adding listeners while shutting down.  Without that protection, what stops some from adding listeners during the shutdown sequence?

Reply:
It was a mistake on my end. I had intended to avoid holding the lock for the entire duration while the sharedInformer ran as per https://github.com/kubernetes/kubernetes/pull/111122#discussion_r930320473. Revisiting this now I realize the lock is only held while the informer is actively stopping. 

I've changed this bit so that the lock is held until the informers all stop.